### ISSUE TITLE
session_t->Create() Failed using tensorflow exported model on IOS

### ISSUE THREAD

**Issue by jakajacky** [tensorflow/tensorflow#12319]
Please go to Stack Overflow for help and support:

https://stackoverflow.com/questions/tagged/tensorflow

If you open a GitHub issue, here is our policy:

1. It must be a bug or a feature request.
2. The form below must be filled out.
3. It shouldn't be a TensorBoard issue. Those go [here](https://github.com/tensorflow/tensorboard/issues).

**Here's why we have that policy**: TensorFlow developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: YES
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: 
Generating Tensorflow model:  Ubuntu 16.04 LTS (GNU/Linux 4.4.0-91-generic x86_64)
Load Tensorflow model: macOS 10.12 Sierra
- **TensorFlow installed from (source or binary)**: TensorFlow is installed from source
- **TensorFlow version (use command below)**: TensorFlow version 1.2.0
- **Python version**: Python version 3.5
- **Bazel version (if compiling from source)**: Bazel version  0.5.3
- **CUDA/cuDNN version**: None
- **GPU model and memory**: NA
- **Exact command to reproduce**:

### Describe the problem
Error happen when call "status = session_t->Create(graph)"

error message:
2017-08-16 13:22:27.692781+0800 CameraExample[6322:2753781] Error adding graph to session: No OpKernel was registered to support Op 'RandomUniform' with these attrs.  Registered devices: [CPU], Registered kernels:
  <no registered kernels>

### Source code / logs
Export models as *.pb file: (In python)
self.saver.save(self.sess, self.my_path_to_model + "model.ckpt", global_step=i)
		output_graph_def = tf.graph_util.convert_variables_to_constants(self.sess,self.sess.graph_def,output_node_names=['models_simple/y_conv'])
			with tf.gfile.FastGFile(self.my_path_to_model+ 'graph.pb', mode = 'wb') as f:
				f.write(output_graph_def.SerializeToString())
------------------------------------------------------------------------------------
IOS:

use the following operations to process exported model
 ------------------------------------------------------------------------------------
#Call freeze_graph, synthesize graph.pb and CKPT, generate frozen.pb
bazel-bin/tensorflow/python/tools/freeze_graph \
  --input_graph=XXXX/graph.pb \
  --input_checkpoint=XXXX/model.ckpt-0 \
  --output_node_names=models_simple/Placeholder,models_simple/y_conv \
  --input_binary \
  --output_graph=XXXX/frozen.pb

#Call optimize_for_inference, reduce the operation, replace the calculation can not run in the iOS version, generate inference.pb
bazel-bin/tensorflow/python/tools/optimize_for_inference \
  --input=XXXX/frozen.pb \
  --output=XXXX/inference.pb \
  --input_names=models_simple/Placeholder \
  --output_names=models_simple/y_conv \
  --frozen_graph=True

#Call quantize_graph to further optimize the graph, generate rounded_graph.pb
bazel-bin/tensorflow/tools/quantization/quantize_graph \
  --input=XXXX/inference.pb \
  --output=XXXX/rounded_graph.pb \
  --output_node_names=models_simple/y_conv \
  --mode=weights_rounded
-------------------------------------------------------------------------------------------

Then, execute the following code, (BOOL)loadGraphFromPath executed successfully, but error raised at "status = session_t->Create(graph)"

- (BOOL)loadGraphFromPath:(NSString *)path
{
  auto status = ReadBinaryProto(tensorflow::Env::Default(), path.fileSystemRepresentation, &graph);
  if (!status.ok()) {
    NSLog(@"Error reading graph: %s", status.error_message().c_str());
    return NO;
  }
  
  // This prints out the names of the nodes in the graph.
  auto nodeCount = graph.node_size();
  NSLog(@"Node count: %d", nodeCount);
  for (auto i = 0; i < nodeCount; ++i) {
    auto node = graph.node(i);
    NSLog(@"Node %d: %s '%s'", i, node.op().c_str(), node.name().c_str());
  }
  
  return YES;
}

create tf session:
- (BOOL)createSession
{
  tensorflow::SessionOptions options;
  auto status = tensorflow::NewSession(options, &session_t);
  if (!status.ok()) {
    NSLog(@"Error creating session: %s", status.error_message().c_str());
    return NO;
  }
  
  status = session_t->Create(graph);
  if (!status.ok()) {
  // error
    NSLog(@"Error adding graph to session: %s", status.error_message().c_str());
    return NO;
  }
  
  return YES;
}

error message:
2017-08-16 13:22:27.692781+0800 CameraExample[6322:2753781] Error adding graph to session: No OpKernel was registered to support Op 'RandomUniform' with these attrs.  Registered devices: [CPU], Registered kernels:
  <no registered kernels>

	 [[Node: dropout/random_uniform/RandomUniform = RandomUniform[T=DT_INT32, dtype=DT_FLOAT, seed=87654321, seed2=24](dropout/Shape)]]

# It seems that RandomUniform was not supported on iOS , but when I called freeze_graph, those which are not supported on iOS should be deleted. What can I do for this problem?

**Comment by skye** [comment]
@petewarden

**Comment by petewarden** [comment]
We'll be updating the main documentation soon, but for now you might find this free guide to freezing and missing variables on mobile useful: http://www.oreilly.com/data/free/building-mobile-applications-with-tensorflow.csp

**Comment by wangyaobupt** [comment]
Peter, thanks for your book. 

jakajacky and I worked in same team, and today we further debugged this issue with help from your guide book, but error still exist.

***Current status is***
1. Freezing model is successful. 

> $ bazel-bin/tensorflow/python/tools/freeze_graph \                       
--input_graph=/Users/xiaoqiang/Desktop/Volcano_Offline_1/graph_dropout.pb \        
--input_checkpoint=/Users/xiaoqiang/Desktop/Volcano_Offline_1/models/model.ckpt-0 \
--input_node_names=models_simple/x --output_node_names=models_simple/y_conv \
--input_binary \
--output_graph=/Users/xiaoqiang/Desktop/Volcano_Offline_1/frozen.pb

> console log:
> 
> 2017-08-17 16:20:37.775619: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.2 instructions, but these are available on your machine and could speed up CPU computations.
> 2017-08-17 16:20:37.775643: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX instructions, but these are available on your machine and could speed up CPU computations.
> 2017-08-17 16:20:37.775650: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX2 instructions, but these are available on your machine and could speed up CPU computations.
> 2017-08-17 16:20:37.775655: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use FMA instructions, but these are available on your machine and could speed up CPU computations.
> Converted 8 variables to const ops.
> 47 ops in the final graph.
> 
> result: 
> frozen.pb

2. Print summary of model, there is a "RandomUniform" op, which will cause error when create session in IOS

> $ bazel run tensorflow/tools/graph_transforms:summarize_graph -- \       
     --in_graph="/Users/xiaoqiang/Desktop/Volcano_Offline_1/inference.pb"

> console log:
> 
> INFO: Found 1 target...
> Target //tensorflow/tools/graph_transforms:summarize_graph up-to-date:
>   bazel-bin/tensorflow/tools/graph_transforms/summarize_graph
> INFO: Elapsed time: 0.173s, Critical Path: 0.00s
> 
> INFO: Running command line: bazel-bin/tensorflow/tools/graph_transforms/summarize_graph '--in_graph=/Users/xiaoqiang/Desktop/Volcano_Offline_1/inference.pb'
> Found 2 possible inputs: (name=models_simple/x, type=float(1), shape=[]) (name=models_simple/Placeholder, type=float(1), shape=[]) 
> No variables spotted.
> Found 1 possible outputs: (name=models_simple/y_conv, op=Softmax) 
> Found 70140 (70.14k) const parameters, 0 (0) variable parameters, and 0 control_edges
> Op types used: 12 Const, 6 Add, 3 Relu, 2 Placeholder, 2 MatMul, 2 MaxPool, 2 Mul, 2 Reshape, 2 Conv2D, 1 RealDiv, 1 RandomUniform, 1 Floor, 1 Shape, 1 Softmax, 1 Sub
> 
 
3. Try "Removing training-only nodes" (bazel run tensorflow/tools/graph_transforms:transform_graph). The operation is successfully completed.

> $ bazel run tensorflow/tools/graph_transforms:transform_graph -- \       
    --in_graph="/Users/xiaoqiang/Desktop/Volcano_Offline_1/inference.pb"  \
    --out_graph="/Users/xiaoqiang/Desktop/Volcano_Offline_1/optimized_inception_graph.pb"  \
    --inputs="models_simple/x" \
    --outputs="models_simple/y_conv" \
    --transforms='strip_unused_nodes(type=float,shape="1,256")'

> console log:
> 
> INFO: Found 1 target...
> Target //tensorflow/tools/graph_transforms:transform_graph up-to-date:
>   bazel-bin/tensorflow/tools/graph_transforms/transform_graph
> INFO: Elapsed time: 0.191s, Critical Path: 0.00s
> 
> INFO: Running command line: bazel-bin/tensorflow/tools/graph_transforms/transform_graph '--in_graph=/Users/xiaoqiang/Desktop/Volcano_Offline_1/inference.pb' '--out_graph=/Users/xiaoqiang/Desktop/Volcano_Offline_1/optimized_inception_graph.pb' '--inputs=models_simple/x' '--outputs=models_simple/y_conv' '--transforms=strip_unused_nodes(type=float,shape="1,256")'
> 2017-08-17 16:27:20.074270: I tensorflow/tools/graph_transforms/transform_graph.cc:262] Applying strip_unused_nodes
> 
> result: 
> optimized_inception_graph.pb

4. Since  "dropout" layers is left in graph, error persist when create session.

> $ bazel run tensorflow/tools/graph_transforms:summarize_graph -- \       
     --in_graph="/Users/xiaoqiang/Desktop/Volcano_Offline_1/optimized_inception_graph.pb"

> console log: 
> 
> INFO: Found 1 target...
> Target //tensorflow/tools/graph_transforms:summarize_graph up-to-date:
>   bazel-bin/tensorflow/tools/graph_transforms/summarize_graph
> INFO: Elapsed time: 0.179s, Critical Path: 0.00s
> 
> INFO: Running command line: bazel-bin/tensorflow/tools/graph_transforms/summarize_graph '--in_graph=/Users/xiaoqiang/Desktop/Volcano_Offline_1/optimized_inception_graph.pb'
> Found 2 possible inputs: (name=models_simple/x, type=float(1), shape=[]) (name=models_simple/Placeholder, type=float(1), shape=[]) 
> No variables spotted.
> Found 1 possible outputs: (name=models_simple/y_conv, op=Softmax) 
> Found 70140 (70.14k) const parameters, 0 (0) variable parameters, and 0 control_edges
> Op types used: 12 Const, 6 Add, 3 Relu, 2 Placeholder, 2 MatMul, 2 MaxPool, 2 Mul, 2 Reshape, 2 Conv2D, 1 RealDiv, 1 RandomUniform, 1 Floor, 1 Shape, 1 Softmax, 1 Sub

***Oher Experiments***
We also tried to remove dropout layer from training code on Ubuntu Server, then re-train model, export model to IOS, the session_t->Create() operation passed. However, we cannot use this approach as solution, because removing dropout layer before training will lead to changes of weights and other parameters in neural networks.

Our objectives are as follow
1. Given a model (Checkpoint + GraphDef) trained and verified on server
2. process the model with tools to remove all unnecessary nodes/ops for inference.
3. inference on IOS

***Question***
How to remove dropout layer from model, since "dropout" is unnecessary in inference procedure.

***Code for model generation (Python)***
`           self.x = tf.placeholder(tf.float32, [None, feature_num], name = 'x')
            x_image = tf.reshape(self.x, [-1, 1, feature_num, 1])
            self.y_ = tf.placeholder(tf.float32, [None, label_dim],name='y_')

            self.W_conv1 = weight_variable([1, self.conv_len, 1, self.num_conv1])
            self.b_conv1 = bias_variable([self.num_conv1])
            self.h_conv1 = tf.nn.relu(conv2d(x_image, self.W_conv1) + self.b_conv1)

            self.h_pool1,pooling_factor = max_pool_1x4(self.h_conv1,pooling_factor)
            self.W_conv2 = weight_variable([1, self.conv_len, self.num_conv1, self.num_conv2])
            self.b_conv2 = bias_variable([self.num_conv2])
            self.h_conv2 = tf.nn.relu(conv2d(self.h_pool1, self.W_conv2) + self.b_conv2)

            self.h_pool2,pooling_factor = max_pool_1x2(self.h_conv2,pooling_factor)
            pooled_size = int(feature_num / pooling_factor * self.num_conv2)
            print ("pooling: (%d,%d) -> %d" % (feature_num, pooling_factor, pooled_size))
            self.W_fc1 = weight_variable([pooled_size, self.num_fullc])
            self.b_fc1 = bias_variable([self.num_fullc])
            self.h_pool2_flat = tf.reshape(self.h_pool2, [-1, pooled_size])
            self.h_fc1 = tf.nn.relu(tf.matmul(self.h_pool2_flat, self.W_fc1) + self.b_fc1)

            self.keep_prob = tf.placeholder(tf.float32)
            self.h_fc1_drop = tf.nn.dropout(self.h_fc1, self.keep_prob)
            self.W_fc2 = weight_variable([self.num_fullc, label_dim])
            self.b_fc2 = bias_variable([label_dim])

            self.y_conv = tf.nn.softmax(tf.matmul(self.h_fc1_drop, self.W_fc2) + self.b_fc2, name='y_conv')

            classes_weights = tf.constant([ 1.0,  0.8,  1.0, 1.0])
            cross = tf.nn.weighted_cross_entropy_with_logits(self.y_,self.y_conv, pos_weight=classes_weights)
            self.cross_entropy = tf.reduce_mean(cross)

            # self.cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(logits=self.y_conv, labels=self.y_))
            self.train_step = tf.train.AdamOptimizer(lr).minimize(self.cross_entropy)`

**Comment by Christ1992** [comment]
Did you solve the problem? I tried to follow the instruction as [https://github.com/JieHe96/iOS_Tensorflow_ObjectDetection_Example](url) to register the ops and recompile. However, the Op registered issues still occurred. If this works for you, or you figure out how to solve it, please let me know asap.

**Comment by jakajacky** [comment]
@Christ1992  Could you provide detailed information about your issues？

**Comment by Christ1992** [comment]
I mainly encountered 2 issues. 
1. "duplicate symbol" issue:
`duplicate symbol __ZN10tensorflow15CreateGPUTracerEv in:
    /Users/XXXX/tensorflow-master/tensorflow/contrib/makefile/gen/lib/libtensorflow-core.a(gpu_tracer.o)
ld: 1 duplicate symbol for architecture arm64`
I solved this by commenting the gpu_tracker.cc line in Makefile. I don't know the side effect of this operation.
2. "No OpKernel was registered to support Op XXX" issue
I solve this by following the above link. I finally recompiled a library that contained all ops of my model, as it listed in my ops_to_registered.h. 

Hope that you can solve your problem soon. =^_^=

**Comment by jakajacky** [comment]
@Christ1992  it's good news^_^. Unfortunately I tried  JieHe‘s instruc, but terminal bring me some issues like this when i take step "compile_ios_tensorflow.sh":
```
1 error generated.
make: *** [/Users/xiaoqiang/2Github/MachineLearning/tensorflow-r1.2/tensorflow/contrib/makefile/gen/obj/ios_ARMV7/tensorflow/core/kernels/dequantize_op.o] Error 1
make: *** Waiting for unfinished jobs....
In file included from tensorflow/core/kernels/quantization_utils.cc:16:
./tensorflow/core/kernels/quantization_utils.h:29:10: fatal error: 'public/gemmlowp.h' file not
      found
#include "public/gemmlowp.h"
```
that's only one of my issues. They are the same kind of problem, and my tensorFlow  have no file or folder like  'public/gemmlowp.h'  indeedly. Maybe you have some ideas

**Comment by Christ1992** [comment]
I have no idea about this. Maybe you could try to compile with the Makefile I used? [Makefile - google drive](https://drive.google.com/file/d/0BxmUvQp8r-wJMXVmVU9HejZpUTA/view?usp=sharing)

**Comment by jakajacky** [comment]
@Christ1992 Thanks for your reply,  i'll try it right away √

**Comment by jakajacky** [comment]
@Christ1992 i saw your new issue #12560, but my issue "No OpKernel was registered to support Op 'RandomUniform' " have been resolved by remove 'dropout' only be used to train model, then i met "No OpKernel was registered to support Op 'Mul'" when run official tensorflow-ios camera demo, finaly i tried freezing the graph, and it worked. In view of #12560, the only difference between you and me is that i have not used 'pods' to install, and DT_FLOAT or INT32 have nothing to do with this issue probably. I can only provide these. v=v

**Comment by WangChengLongWilliam** [comment]
sorry i can't understand your meaning,i guess this question is not about my project

At 2017-08-25 10:33:41, "X-Roach" <notifications@github.com> wrote:


@Christ1992 i saw your new issue #12560, but my issue "No OpKernel was registered to support Op 'RandomUniform' " have been resolved by remove 'dropout' only be used to train model, then i met "No OpKernel was registered to support Op 'Mul'" when run official tensorflow-ios camera demo, finaly i tried freezing the graph, and it worked. In view of #12560, the only difference between you and me is that i have not used 'pods' to install, and DT_FLOAT or INT32 have nothing to do with this issue probably. I can only provide these. v=v

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.
